### ISSUE TITLE
Consider traits for STRUCT with named fields.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp-spanner#592]
Cloud Spanner supports STRUCT with named fields (e.g. `STRUCT<FirstName STRING, LastName STRING>`) or with anonymous fields (`STRUCT<STRING, STRING>`).

The latter is represented by `std::tuple<std::string, std::string>`.  The former is (currently) represented by `std::tuple<std::pair<std::string, std::string>, std::pair<std::string, std::string>>`, and it requires that the names of the fields are provided in each object.

An alternative design would be to require a tuple-like C++ `struct` for spanner `STRUCT` with named fields, where the application provide a trait and/or accessor for the field names.  For example (using C++17 syntax for exposition):

```C++
struct NamedStructViaMembers {
  std::int64_t id;
  std::string first_name;
  std::string last_name;

   template <std::size_t N>
  constexpr char const* get_element_name() const {
    static_assert(N <= 2);
    if constexpr (N == 0) {
      return "id";
    }
    if constexpr (N == 1) {
      return "FirstName";
    }
    if constexpr (N == 2) {
      return "LastName";
    }
  }
};
```

For C++11, the application would need to specialize `get_element_name()` for `N={0,1,2}`

In *addition* the application would need to define the `get<I>()` accessors, like for any tuple-like struct.

**Comment by coryan** [comment]
A prototype of this change is available at:

https://github.com/googleapis/google-cloud-cpp-spanner/compare/master...coryan:prototype-named-structs?expand=1

I am inclined to say that we should **not** implement this idea.  The code looks neat at the point of invocation, but it is a lot of boilerplate for C++11, and for C++17 it is *also* a lot of boilerplate.

**Comment by devjgm** [comment]
We should note that even today a user could define a constructor on their custom struct from the tuple they get from the query. Something like:

```cc
struct NamedStruct {
  std::int64_t id;
  std::string first_name;
  std::string last_name;

  // Handles mapping from the tuple fields to their struct fields
  NamedStruct(std::tuple<std::int64_t, std::string, std::string> tup)
    : id(std::get<0>(tup)), first_name(std::get<1>(tup)), last_name(std::get<2>(tup)) {}
};

...
Value v = ...
auto tup = v.get<std::tuple<std::int64_t, std::string, std::string>>();
if (tup) {
  NamedStruct ns(*tup);
}
```

**Comment by devjgm** [comment]
FTR, I'm not sure that I'm totally clear on what part of the API needs to be simplified (potentially). Creating structs with named fields for use in, say, inserts or SQL query params? Or extracting named structs out of `Value` objects that resulted form a query? Or maybe both?

My concern is that I'm not sure we understand the real use cases well enough to design a simpler solution right now.

Maybe a simpler solution involves `std::vector`, or `std::map` holding type-erased `Value` objects. Or maybe it involves a "tuple builder" that lets users specify struct field names once (to the builder's c'tor), but then it knows how to build structs-with-named-fields from other tuples without names? Or maybe a meta-programmed solution that inspects the fields in a `struct`, similar to @coryan's proposal? Without knowing a bit more about actual use cases and pain points, I'm unsure how to compare these various options.

So, honestly, I'm sorta inclined to do nothing on this issue at this point, until we know more clearly where the pain.

I'm curious if others have thoughts on this issue.

**Comment by coryan** [comment]
> My concern is that I'm not sure we understand the real use cases well enough to design a simpler solution right now.

I think we completely agree on this one, and I agree with you that *for now* we are probably better off doing nothing (maybe kicking this issue to the GA milestone or simply "no milestone").

> I'm curious if others have thoughts on this issue.

I think I filed this when writing the simplest possible example of a query with a named struct parameter value.  I found it very repetitive to pass the names over and over.  I was able to write a helper, but I wonder if we can do better, by separating the type that holds the values the traits about the value names.

In any case: move it out of the milestone, we can probably add additional APIs when we figure out how to do this better.
